File: observability/instrumentation/node/observability.ts
/**
 * OFFCON observability for Node (Fastify) services.
 *
 * Usage in a service's server.ts:
 *
 *   import { setupObservability, businessMetrics } from "@offcon/observability";
 *
 *   const app = Fastify();
 *   await setupObservability(app, { serviceName: "payment-svc", tier: "billing" });
 *   businessMetrics.paymentAttempts.inc({ result: "succeeded" });
 *
 * Provides Prometheus /metrics (RED metrics with route templating),
 * OpenTelemetry tracing via OTLP, and business metric helpers. Names match the
 * conventions in prometheus/prometheus.yml.
 */

import type { FastifyInstance, FastifyReply, FastifyRequest } from "fastify";
import {
  Counter,
  Gauge,
  Histogram,
  Registry,
  collectDefaultMetrics,
} from "prom-client";
import { NodeSDK } from "@opentelemetry/sdk-node";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-grpc";
import { Resource } from "@opentelemetry/resources";
import {
  ATTR_SERVICE_NAME,
  ATTR_DEPLOYMENT_ENVIRONMENT,
} from "@opentelemetry/semantic-conventions/incubating";
import { getNodeAutoInstrumentations } from "@opentelemetry/auto-instrumentations-node";
import {
  ParentBasedSampler,
  TraceIdRatioBasedSampler,
} from "@opentelemetry/sdk-trace-node";

export const registry = new Registry();

const LATENCY_BUCKETS = [0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5];

export interface ObservabilityOptions {
  serviceName: string;
  tier?: string;
  otlpEndpoint?: string;
  sampleRatio?: number;
}

/* -------------------------------------------------------------------------- */
/* RED metrics                                                                */
/* -------------------------------------------------------------------------- */
let httpRequests: Counter<string>;
let httpDuration: Histogram<string>;
let httpInflight: Gauge<string>;
let serviceName = "unknown";

let sdk: NodeSDK | undefined;

export async function setupObservability(
  app: FastifyInstance,
  opts: ObservabilityOptions,
): Promise<void> {
  serviceName = opts.serviceName;
  const tier = opts.tier ?? "unknown";
  const endpoint =
    opts.otlpEndpoint ??
    process.env.OTEL_EXPORTER_OTLP_ENDPOINT ??
    "http://otel-collector:4317";
  const sampleRatio = opts.sampleRatio ?? 1.0;

  // ---- Tracing ----
  sdk = new NodeSDK({
    resource: new Resource({
      [ATTR_SERVICE_NAME]: serviceName,
      tier,
      [ATTR_DEPLOYMENT_ENVIRONMENT]: process.env.DEPLOY_ENV ?? "production",
    }),
    traceExporter: new OTLPTraceExporter({ url: endpoint }),
    sampler: new ParentBasedSampler({
      root: new TraceIdRatioBasedSampler(sampleRatio),
    }),
    instrumentations: [
      getNodeAutoInstrumentations({
        // fs instrumentation is noisy; disable it
        "@opentelemetry/instrumentation-fs": { enabled: false },
      }),
    ],
  });
  sdk.start();

  // ---- Default + RED metrics ----
  collectDefaultMetrics({ register: registry, prefix: "" });

  httpRequests = new Counter({
    name: "http_requests_total",
    help: "Total HTTP requests.",
    labelNames: ["service", "method", "path", "status"],
    registers: [registry],
  });
  httpDuration = new Histogram({
    name: "http_request_duration_seconds",
    help: "HTTP request latency.",
    labelNames: ["service", "method", "path", "status"],
    buckets: LATENCY_BUCKETS,
    registers: [registry],
  });
  httpInflight = new Gauge({
    name: "http_inflight_requests",
    help: "In-flight HTTP requests.",
    labelNames: ["service"],
    registers: [registry],
  });

  // ---- Per-request RED hooks ----
  app.addHook("onRequest", async (req: FastifyRequest) => {
    (req as RequestWithTiming)._startTime = process.hrtime.bigint();
    httpInflight.inc({ service: serviceName });
  });

  app.addHook("onResponse", async (req: FastifyRequest, reply: FastifyReply) => {
    httpInflight.dec({ service: serviceName });
    const start = (req as RequestWithTiming)._startTime;
    const elapsed = start ? Number(process.hrtime.bigint() - start) / 1e9 : 0;
    // Fastify exposes the matched route template at req.routeOptions.url
    const path = req.routeOptions?.url ?? "unmatched";
    const labels = {
      service: serviceName,
      method: req.method,
      path,
      status: String(reply.statusCode),
    };
    httpRequests.inc(labels);
    httpDuration.observe(labels, elapsed);
  });

  // ---- Endpoints ----
  app.get("/metrics", async (_req, reply) => {
    reply.header("Content-Type", registry.contentType);
    return registry.metrics();
  });

  app.get("/healthz", async () => ({ status: "ok", service: serviceName }));

  // Graceful shutdown of the trace SDK
  app.addHook("onClose", async () => {
    await sdk?.shutdown();
  });
}

interface RequestWithTiming {
  _startTime?: bigint;
}

File: observability/instrumentation/go/observability.go
// Package observability provides drop-in metrics + tracing + structured
// logging for OFFCON Go services. A service wires it up with a single call:
//
//	shutdown, err := observability.Init(ctx, observability.Config{
//	    ServiceName: "auth-svc",
//	    Tier:        "core",
//	})
//	defer shutdown(ctx)
//
// It exposes a Prometheus /metrics handler, instruments HTTP + gRPC with RED
// metrics, and exports OTLP traces to the collector. All metric names follow
// the conventions documented in prometheus/prometheus.yml.
package observability

import (
	"context"
	"fmt"
	"net/http"
	"os"

	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promhttp"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
	"go.opentelemetry.io/otel/trace"
)

// Config configures the observability stack for a service.
type Config struct {
	ServiceName string
	Tier        string
	// OTLPEndpoint defaults to the OTEL_EXPORTER_OTLP_ENDPOINT env or
	// "otel-collector:4317".
	OTLPEndpoint string
	// SampleRatio is the parent-based sample ratio when the collector isn't
	// doing tail sampling. Defaults to 1.0 (collector decides).
	SampleRatio float64
}

// Metrics holds the standard RED metric collectors for a service.
type Metrics struct {
	HTTPRequests *prometheus.CounterVec
	HTTPDuration *prometheus.HistogramVec
	HTTPInflight prometheus.Gauge
	GRPCHandled  *prometheus.CounterVec
	GRPCDuration *prometheus.HistogramVec
	registry     *prometheus.Registry
}

var (
	globalMetrics *Metrics
	tracer        trace.Tracer
)

// Init sets up metrics + tracing and returns a shutdown function.
func Init(ctx context.Context, cfg Config) (func(context.Context) error, error) {
	if cfg.ServiceName == "" {
		return nil, fmt.Errorf("observability: ServiceName is required")
	}
	if cfg.OTLPEndpoint == "" {
		cfg.OTLPEndpoint = envOr("OTEL_EXPORTER_OTLP_ENDPOINT", "otel-collector:4317")
	}
	if cfg.SampleRatio == 0 {
		cfg.SampleRatio = 1.0
	}

	globalMetrics = newMetrics(cfg)

	// Tracing: OTLP gRPC exporter → collector
	exp, err := otlptracegrpc.New(ctx,
		otlptracegrpc.WithEndpoint(cfg.OTLPEndpoint),
		otlptracegrpc.WithInsecure(),
	)
	if err != nil {
		return nil, fmt.Errorf("observability: trace exporter: %w", err)
	}

	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName(cfg.ServiceName),
			attribute.String("tier", cfg.Tier),
			semconv.DeploymentEnvironment(envOr("DEPLOY_ENV", "production")),
		),
	)
	if err != nil {
		return nil, fmt.Errorf("observability: resource: %w", err)
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exp),
		sdktrace.WithResource(res),
		sdktrace.WithSampler(sdktrace.ParentBased(sdktrace.TraceIDRatioBased(cfg.SampleRatio))),
	)
	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{}, propagation.Baggage{},
	))
	tracer = tp.Tracer(cfg.ServiceName)

	return func(shutdownCtx context.Context) error {
		return tp.Shutdown(shutdownCtx)
	}, nil
}

// MetricsHandler returns the Prometheus /metrics HTTP handler.
func MetricsHandler() http.Handler {
	if globalMetrics == nil {
		return promhttp.Handler()
	}
	return promhttp.HandlerFor(globalMetrics.registry, promhttp.HandlerOpts{})
}

// Tracer returns the service tracer (safe even before Init — returns a noop).
func Tracer() trace.Tracer {
	if tracer == nil {
		return otel.Tracer("uninitialized")
	}
	return tracer
}

func newMetrics(cfg Config) *Metrics {
	reg := prometheus.NewRegistry()
	// Standard process + Go runtime collectors
	reg.MustRegister(prometheus.NewGoCollector())
	reg.MustRegister(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}))

	constLabels := prometheus.Labels{"service": cfg.ServiceName}

	m := &Metrics{
		registry: reg,
		HTTPRequests: prometheus.NewCounterVec(prometheus.CounterOpts{
			Name: "http_requests_total", Help: "Total HTTP requests.", ConstLabels: constLabels,
		}, []string{"method", "path", "status"}),
		HTTPDuration: prometheus.NewHistogramVec(prometheus.HistogramOpts{
			Name: "http_request_duration_seconds", Help: "HTTP request latency.", ConstLabels: constLabels,
			Buckets: []float64{0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5},
		}, []string{"method", "path", "status"}),
		HTTPInflight: prometheus.NewGauge(prometheus.GaugeOpts{
			Name: "http_inflight_requests", Help: "In-flight HTTP requests.", ConstLabels: constLabels,
		}),
		GRPCHandled: prometheus.NewCounterVec(prometheus.CounterOpts{
			Name: "grpc_server_handled_total", Help: "Total gRPC calls handled.", ConstLabels: constLabels,
		}, []string{"grpc_service", "grpc_method", "grpc_code"}),
		GRPCDuration: prometheus.NewHistogramVec(prometheus.HistogramOpts{
			Name: "grpc_server_handling_seconds", Help: "gRPC handling latency.", ConstLabels: constLabels,
			Buckets: []float64{0.002, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1},
		}, []string{"grpc_service", "grpc_method"}),
	}
	reg.MustRegister(m.HTTPRequests, m.HTTPDuration, m.HTTPInflight, m.GRPCHandled, m.GRPCDuration)
	return m
}

func envOr(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}
